File: python/src/kclone/resource.py
from __future__ import annotations

from typing import Dict, List

from .state import ClusterState


def node_resource_table(state: ClusterState) -> List[Dict[str, int]]:
    rows = []
    for node in state.nodes.values():
        rows.append(
            {
                "name": node.name,
                "cpu_used": node.cpu_allocated,
                "cpu_capacity": node.cpu_capacity,
                "mem_used": node.mem_allocated,
                "mem_capacity": node.mem_capacity,
            }
        )
    return rows


def cluster_capacity(state: ClusterState) -> Dict[str, int]:
    total_cpu = sum(n.cpu_capacity for n in state.nodes.values())
    total_mem = sum(n.mem_capacity for n in state.nodes.values())
    used_cpu = sum(n.cpu_allocated for n in state.nodes.values())
    used_mem = sum(n.mem_allocated for n in state.nodes.values())
    return {
        "cpu_used": used_cpu,
        "cpu_total": total_cpu,
        "mem_used": used_mem,
        "mem_total": total_mem,
    }

File: go/pkg/resource/resource.go
package resource

import (
	"kclone-go/pkg/state"
)

// NodeResourceInfo represents resource info for a node
type NodeResourceInfo struct {
	Name        string
	CPUUsed     int
	CPUCapacity int
	MemUsed     int
	MemCapacity int
}

// ClusterCapacity represents total cluster resources
type ClusterCapacity struct {
	CPUUsed  int
	CPUTotal int
	MemUsed  int
	MemTotal int
}

// GetNodeResourceTable returns resource info for all nodes
func GetNodeResourceTable(cs *state.ClusterState) []NodeResourceInfo {
	nodes := cs.GetNodes()
	result := make([]NodeResourceInfo, 0, len(nodes))

	for _, node := range nodes {
		result = append(result, NodeResourceInfo{
			Name:        node.Name,
			CPUUsed:     node.CPUAllocated,
			CPUCapacity: node.CPUCapacity,
			MemUsed:     node.MemAllocated,
			MemCapacity: node.MemCapacity,
		})
	}

	return result
}

// GetClusterCapacity returns total cluster resource usage
func GetClusterCapacity(cs *state.ClusterState) ClusterCapacity {
	nodes := cs.GetNodes()

	capacity := ClusterCapacity{}
	for _, node := range nodes {
		capacity.CPUTotal += node.CPUCapacity
		capacity.MemTotal += node.MemCapacity
		capacity.CPUUsed += node.CPUAllocated
		capacity.MemUsed += node.MemAllocated
	}

	return capacity
}
